File: lib/items/objects.rb
require_relative 'base'
require_relative 'info/info'

OBJECT_FIELDS = {
  base: [
    { key: :type,          type: :int },
    { key: :name,          type: :name },
    { key: :parent_folder, type: :int }
  ],
  animation: [
    { key: :name, type: :name },
    { key: :unknown1,          type: :int   },
    { key: :unknown2,          type: :float },
    { key: :unknown3,          type: :float },
    { key: :unknown4,          type: :int   },
    { key: :alway_negative100, type: :float },
    { key: :unknown5,          type: :bool  },
    { key: :unknown6,          type: :bool  }
  ],
  animation_unknown: [
    { key: :name, type: :name },
    { key: :unknown1, type: :float }

  ]
}.freeze

module Wld
  module Items
    class Objects < Base
      def initialize(file_or_hash, marker = 'OBJS')
        @separator = 'OBJ '
        super
      end

      protected

      def unpack_node(file, index)
        item = { index: index + 1 }

        file.read OBJECT_FIELDS[:base], item
        item[:animations] = unpack_animations(file)
        item[:info] = Info::Info.new(file).to_hash

        item
      end

      def pack_node(node)
        bindata = BinaryDataBuffer.new
        bindata.add OBJECT_FIELDS[:base], node
        bindata.concat pack_animations(node[:animations])
        bindata.concat Info::Info.new(node[:info]).to_binary

        bindata.data
      end

      def unpack_animations(file)
        count_of_animations = file.int
        count_of_animations.times.map do
          animation_item = {}
          file.read OBJECT_FIELDS[:animation], animation_item
          count_of_unknown7 = file.int
          animation_item[:unknown7] = count_of_unknown7.times.map do
            file.read OBJECT_FIELDS[:animation_unknown]
          end
          animation_item
        end
      end

      def pack_animations(animations)
        bindata = BinaryDataBuffer.new

        bindata.push_int animations.size
        animations.each do |animation_item|
          bindata.add OBJECT_FIELDS[:animation], animation_item
          bindata.push_int animation_item[:unknown7].size
          animation_item[:unknown7].each do |unknown|
            bindata.add OBJECT_FIELDS[:animation_unknown], unknown
          end
        end
        bindata.data
      end
    end
  end
end

File: unity_scripts/Assets/Editor/WldImporter/Items/Objects.cs
using System.Collections.Generic;
using WldParser.Helpers;

namespace WldParser.Items
{
    public class Objects : WldSectionBase<ObjectData>
    {
        public Objects(BinaryFileReader file) : base(file, "OBJS", "OBJ ") { }

        protected override ObjectData UnpackNode(BinaryFileReader file, int index, int entrySize)
        {
            var obj = new ObjectData { index = index + 1 };

            // Base
            obj.type = file.Int();
            obj.name = file.Name();
            obj.parent_folder = file.Int();

            // Animations
            obj.animations = UnpackAnimations(file);

            // Info (Nested Parser)
            obj.info = new Info.InfoParser(file).Parse();

            return obj;
        }

        private List<AnimationData> UnpackAnimations(BinaryFileReader file)
        {
            int count = file.Int();
            var list = new List<AnimationData>();
            for (int i = 0; i < count; i++)
            {
                var anim = new AnimationData();
                anim.name = file.Name();
                anim.model_3d_id = file.Int();
                anim.unknown2 = file.Float();
                anim.unknown3 = file.Float();
                anim.unknown4 = file.Int();
                anim.always_negative100 = file.Float();
                anim.loop_animation = file.Bool();
                anim.default_action = file.Bool();

                int u7Count = file.Int();
                for (int j = 0; j < u7Count; j++)
                {
                    anim.unknown7.Add(new AnimationUnknown 
                    { 
                        name = file.Name(), 
                        unknown1 = file.Float() 
                    });
                }
                list.Add(anim);
            }
            return list;
        }
    }
}